NotificationBadge added to navbar

vrouilhac opened the pull request:
<!---
Please describe the pull request below.
If it closes an issue, make sure to write "closes #1234"
If there is an issue but it isn't completely closed, still refer to the issue number, eg. "part of #1234"
--->


<!---
Checklists - you can remove one that is not applicable (ie. remove backend checklist if you only worked on frontend)
If you need help with any of these, please ask :)
--->
**Backend checklist**
- [ ] Formatted my code by running `isort . && black .` in `app/backend`
- [ ] Added tests for any new code or added a regression test if fixing a bug
- [ ] All tests pass
- [ ] Run the backend locally and it works
- [ ] Added migrations if there are any database changes, rebased onto `develop` if necessary for linear migration history

**Frontend checklist**
- [ ] Formatted my code with `yarn format && yarn lint --fix`
- [ ] There are no warnings from `yarn lint`
- [ ] There are no console warnings when running the app
- [ ] Added any new components to storybook
- [ ] Added tests where relevant
- [ ] All tests pass
- [ ] Clicked around my changes running locally and it works
- [ ] Checked Desktop, Mobile and Tablet screen sizes


<!---
Remember to request review from couchers-org/frontend, couchers-org/@backend or an individual.
Once your code is approved, remember to merge it if you have write access
--->

Commit:
NotificationBadge added to navbar

vrouilhac commented:
I added a notification badge so when we receive a message we can see it on the navigation bar even if we are on other tabs than messages. The issue though is that on desktop the badge override the right button as you can see with the pictures. I can add a prop to NotificationBadge as optional to place it a little upper to solve this ?
![02](https://user-images.githubusercontent.com/62873665/120078856-b1d1e780-c0b1-11eb-82ab-415c2843095f.png)
![01](https://user-images.githubusercontent.com/62873665/120078863-b5656e80-c0b1-11eb-9948-4c7419c57539.png)

vrouilhac commented:
I think reducing the size of the badge on desktop should be a good idea here's what I get 
![03](https://user-images.githubusercontent.com/62873665/120079222-396c2600-c0b3-11eb-8068-1db05bae63ee.png)
 For this I added a prop to NavigationBadge named 'up' optional and boolean that is set to true for desktop version

aapeliv commented on app/frontend/src/components/Navigation/Navigation.tsx:
```diff
@@ -10,8 +10,10 @@ import {
 import classNames from "classnames";
 import { CloseIcon, MenuIcon } from "components/Icons";
 import ExternalNavButton from "components/Navigation/ExternalNavButton";
+import NotificationBadge from "components/NotificationBadge";
 import { useAuthContext } from "features/auth/AuthProvider";
 import useAuthStyles from "features/auth/useAuthStyles";
+import useNotifications from "features/useNotifications";
```
I think you forgot to add these to git!

vrouilhac replied:
You mean those lines wasn't added to the commit ? @aapeliv

aapeliv replied:
Sorry I'm mistaken here, please ignore the comment! I didn't realize these files were already in the tree, I though they were files you added.

aapeliv reviewed (commented):
I think you forgot to add a few files to git; also there's an extra `.Navigation.tsx.un~` file that shouldn't be added!

vrouilhac commented:
> I think you forgot to add a few files to git; also there's an extra `.Navigation.tsx.un~` file that shouldn't be added!

I'll remove those

lucaslcode commented on app/frontend/src/components/Navigation/Navigation.tsx:
```diff
@@ -115,13 +117,22 @@ export default function Navigation() {
   const [open, setOpen] = React.useState(false);
+  const { data } = useNotifications();
+    
+  let notificationNumber = data?.unseenMessageCount;
```
Make this `const` and also make it the sum of the unread messages and also host requests, since they are under this page too.

lucaslcode reviewed (changes requested):
In `NotificationBade.tsx`, the styling only really works in the place where it was used. Try changing it to this:
```
const useStyles = makeStyles((theme) => ({
  badge: {
    right: "-0.5rem", //changed from "theme.spacing(3)" because it's too far for the general case, and also it isn't related to spacing, just offsetting a specific amount that isn't spacing-related, so we should use rem directly
    //delete: top: "50%",
  },
}));
```

lucaslcode commented on app/frontend/src/components/Navigation/Navigation.tsx:
```diff
@@ -115,13 +117,22 @@ export default function Navigation() {
         {menu.map(({ name, route }) => (
           <ListItem button key={name}>
-            <NavButton route={route} label={name} labelVariant="h2" />
+            {route === messagesRoute ? (
+              <NotificationBadge count={notificationNumber}>
+                <NavButton route={route} label={name} labelVariant="h2" />
+              </NotificationBadge>
+            ) : (
+              <NavButton route={route} label={name} labelVariant="h2" />
+            )}
```
This doesn't seem ideal but I can't think of a better way. Later we will have a dedicated notifications dropdown like facebook so I think it's fine for now.

vrouilhac replied:
Dedicated notifications seems a good idea, will the message tab be remaining in the navbar then?

vrouilhac commented:
> In `NotificationBade.tsx`, the styling only really works in the place where it was used. Try changing it to this:
> 
> ```
> const useStyles = makeStyles((theme) => ({
>   badge: {
>     right: "-0.5rem", //changed from "theme.spacing(3)" because it's too far for the general case, and also it isn't related to spacing, just offsetting a specific amount that isn't spacing-related, so we should use rem directly
>     //delete: top: "50%",
>   },
> }));
> ```

This still seem to be on top of search button, because of the padding of the message NavButton i guess

Commit:
NavButton style changed

vrouilhac commented:
Here is the result 
-0.7rem : 
![03](https://user-images.githubusercontent.com/62873665/120101945-01172700-c149-11eb-842b-5c62b01832e2.png)

-0.5rem :
![02](https://user-images.githubusercontent.com/62873665/120101950-0a07f880-c149-11eb-9659-4b4d14891e0b.png)

lucaslcode commented:
Hmm. How about this. In `Navigation.tsx`, change the menu definition at the top to be:
```
const menu = (data: ReturnType<typeof useNotifications>["data"]) => [
  ...
  {
    name: "Messages",
    route: messagesRoute,
    notificationCount: data.whatever,
  },
  ...
```

Then you can do `menu(data).map(...`, and then pass `notificationCount` to `NavButton`, and put the badge in there (it's optional) somewhere that can avoid the padding.

Does that make sense? That will also solve the part I said isn't ideal.

PS. remember to run `yarn lint --fix` to sort the imports, or the CI won't build the preview

vrouilhac commented:
> Hmm. How about this. In `Navigation.tsx`, change the menu definition at the top to be:
> 
> ```
> const menu = (data: ReturnType<typeof useNotifications>["data"]) => [
>   ...
>   {
>     name: "Messages",
>     route: messagesRoute,
>     notificationCount: data.whatever,
>   },
>   ...
> ```
> 
> Then you can do `menu(data).map(...`, and then pass `notificationCount` to `NavButton`, and put the badge in there (it's optional) somewhere that can avoid the padding.
> 
> Does that make sense? That will also solve the part I said isn't ideal.
> 
> PS. remember to run `yarn lint --fix` to sort the imports, or the CI won't build the preview

Yeah look definitely much better but it think we should add right spacing to the message button. 

![01](https://user-images.githubusercontent.com/62873665/120103495-252a3680-c150-11eb-85fb-ae3085b2d95c.png)

( I also added restructuring because we were using .map((item) .., changed to .map(({ route, name, notificationCount })... )

vrouilhac commented:
I tried to add marginLeft to the label ( inside useNavLinkStyle ) which move a bit to the left and help keep the badge tied to the message button, tell me if this is a good way to solve this I'm not sure about the overall impact.

![01](https://user-images.githubusercontent.com/62873665/120103640-ec3e9180-c150-11eb-82ad-c156c9aa5636.png)
 
PS : the marginLeft is set to 0.5rem

Commit:
spacing added to useNavLinkStyles

lucaslcode commented:
How does it look adding that margin to both sides, and only if the notification badge is shown (if it makes a difference)?

Commit:
conditionnal spacing

vrouilhac commented:
I don't know if this is a good way of doing so but I conditionally added spacing ( `className={[class1, condition ? class2 : null].join(' ')}` ), it doesn't change that much from the previous version but unnecessary space was removed.

En result :
![01](https://user-images.githubusercontent.com/62873665/120107716-6b889100-c162-11eb-8cdc-e7511a635113.png)

PS : `class2` contains margin only if the badge is present ( notificationCount exists )

lucaslcode commented on app/frontend/src/components/Navigation/NavButton.tsx:
```diff
@@ -8,24 +9,31 @@ interface NavButtonProps {
       exact={route === baseRoute || route === userRoute}
       to={{ pathname: route }}
-      className={classes.link}
+      className={[
+        classes.link,
+        notificationCount ? classes.notification : null,
+      ].join(" ")}
```
```suggestion
      className={classNames(classes.link, [classes.notification]: !!notificationCount},
```
And import `classNames`

vrouilhac replied:
I did not know that classnames was installed, I added it to the last commit. Also I think I get what !!value do but I struggle to understand it's deepest implementation, I mean how it really works.

lucaslcode replied:
It's the same as `Boolean(value)`, converts to a Boolean. 

If value is truthy (for a number, not zero), !value gives false, so !!value gives true. And the reverse for a falsey value.

lucaslcode replied:
You don't need to do this conversion in `if (value)`, but in some places, you have to have either true or false, not just truthy or falsey

lucaslcode reviewed (changes requested):
Looks good. See below for correct way to do multiple/conditional class names

Commit:
use of classNames
